### ISSUE TITLE
PAM authentication for nginx not working on NixOS

### ISSUE THREAD

**Issue by jluttine** [NixOS/nixpkgs#117578]
**Describe the bug**

Using PAM authentication with nginx doesn't work. The credentials aren't accepted and just asked again (and again).

**To Reproduce**

Add the following to you `configuration.nix` (based on https://nixos.wiki/wiki/Nginx):

```
  security.pam.services.nginx.setEnvironment = false;
  systemd.services.nginx.serviceConfig.SupplementaryGroups = [ "shadow" ];
  services.nginx = {
    enable = true;
    package = (pkgs.nginx.override { modules = [ pkgs.nginxModules.pam ]; });
    virtualHosts.foo = {
      root = "/var/www";
      extraConfig = ''
        auth_pam  "Foo Bar";
        auth_pam_service_name "nginx";
      '';
    };
  };
```

Create `/var/lib/www/index.html` with some content:

```
Hello World!
```

Make sure the file is readable by nginx:

```
$ ls -la /var/lib/www
total 12
drwxr-xr-x  2 jluttine users 4096 Mar 25 12:40 .
drwxr-xr-x 42 root     root  4096 Mar 25 11:32 ..
-rw-r--r--  1 jluttine users   13 Mar 25 11:32 index.html
```
Also, make sure that `/etc/shadow` is readable by group `shadow` (this was fixed very recently in nixpkgs):

```
$ ls -l /etc/shadow
-rw-r----- 1 root shadow 1029 Mar 25 12:27 /etc/shadow
```

Run, for instance, `sudo nixos-rebuild test`, and browse to `127.0.0.1`. The web page asks for authentication. Enter your username and password.

The authentication fails and credentials are asked again (and again and again ...).

From `journalctl -u nginx`:

```
Mar 25 12:32:03 leevi [1012]: pam_warn(nginx:auth): function=[pam_sm_authenticate] flags=0x1 service=[nginx] terminal=[<unknown>] user=[jluttine] ruser=[<unknown>] rhost=[127.0.0.1]
Mar 25 12:32:03 leevi nginx[1012]: 2021/03/25 12:32:03 [error] 1012#1012: *2 PAM: user 'jluttine' - not authenticated: Authentication failure, client: 127.0.0.1, server: foo, request: "GET / HTTP/1.1", host: "127.0.0.1"
```

**Expected behavior**

The authentication should pass and you should see "Hello World!" page.

**Metadata**
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.4.106, NixOS, 21.05.git.f5e8bdd07d1 (Okapi)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.3.10`
 - nixpkgs: `/etc/nixpkgs`
```

Maintainer information:
```yaml
# a list of nixpkgs attributes affected by the problem
attribute:
# a list of nixos modules affected by the problem
module: services.nginx
```

**Comment by jluttine** [comment]
I tried to debug with `pamtester`, but I wasn't able to reproduce the issue. If needed, install `pamtester` (e.g., `nix-shell -p pamtester`). Then (replace `jluttine` with your username):

```
$ sudo -u nginx pamtester nginx jluttine authenticate
Password:
pamtester: successfully authenticated
```

So, this works, but not when trying to authenticate via web browser..

**Comment by jluttine** [comment]
I asked about this also upstream: https://github.com/linux-pam/linux-pam/issues/340

**Comment by jluttine** [comment]
Interestingly, if I run nginx from the terminal without systemd, then it works:

```
$ sudo mkdir /run/nginx
$ sudo chown nginx /run/nginx
$ sudo -u nginx /nix/store/yjqmikwbd06b0yqi11yf9zv4p3jhz61h-nginx-1.18.0/bin/nginx -c /nix/store/1r51wckcm33lwznycmfxb1z8m1faf3va-nginx.conf
```

Just replace the nix store paths (those can be checked from the output of `systemctl status nginx`). Also, I needed to set the port to 8000 so I can run it like this.

```
    ...
    virtualHosts.foo = {
      ...
      listen = [
        {
          addr = "127.0.0.1";
          port = 8000;
          ssl = false;
        }
      ];
     ...
    };
    ...
```

So, I suppose it has something to do with systemd then? Any ideas how to debug further?

**Comment by jluttine** [comment]
I'm now guessing it's related to the security configuration of nginx systemd service: https://github.com/NixOS/nixpkgs/blob/d3f7e969b9860fb80750147aeb56dab1c730e756/nixos/modules/services/web-servers/nginx/default.nix#L771

```
      serviceConfig = {
        ExecStart = execCommand;
        ExecReload = [
          "${execCommand} -t"
          "${pkgs.coreutils}/bin/kill -HUP $MAINPID"
        ];
        Restart = "always";
        RestartSec = "10s";
        # User and group
        User = cfg.user;
        Group = cfg.group;
        # Runtime directory and mode
        RuntimeDirectory = "nginx";
        RuntimeDirectoryMode = "0750";
        # Cache directory and mode
        CacheDirectory = "nginx";
        CacheDirectoryMode = "0750";
        # Logs directory and mode
        LogsDirectory = "nginx";
        LogsDirectoryMode = "0750";
        # Capabilities
        AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
        CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
        # Security
        NoNewPrivileges = true;
        # Sandboxing
        ProtectSystem = "strict";
        ProtectHome = mkDefault true;
        PrivateTmp = true;
        PrivateDevices = true;
        ProtectHostname = true;
        ProtectKernelTunables = true;
        ProtectKernelModules = true;
        ProtectControlGroups = true;
        RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
        LockPersonality = true;
        MemoryDenyWriteExecute = !(builtins.any (mod: (mod.allowMemoryWriteExecute or false)) (optionals (cfg.package ? modules) cfg.package.modules));
        RestrictRealtime = true;
        RestrictSUIDSGID = true;
        PrivateMounts = true;
        # System Call Filtering
        SystemCallArchitectures = "native";
      };
```

**Comment by jluttine** [comment]
Tested disabling those systemd config options one-by-one. This the "maximal" configuration that didn't have the issue:

```
serviceConfig = {
  ExecStart = execCommand;
  ExecReload = [
    "${execCommand} -t"
    "${pkgs.coreutils}/bin/kill -HUP $MAINPID"
  ];
  Restart = "always";
  RestartSec = "10s";
  User = cfg.user;
  Group = cfg.group;
  RuntimeDirectory = "nginx";
  RuntimeDirectoryMode = "0750";
  CacheDirectory = "nginx";
  CacheDirectoryMode = "0750";
  LogsDirectory = "nginx";
  LogsDirectoryMode = "0750";
  AmbientCapabilities = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
  CapabilityBoundingSet = [ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
  ProtectSystem = "strict";  # OK
  ProtectHome = mkDefault true;  # OK
  PrivateTmp = true;  # OK
  ProtectControlGroups = true;  # OK
  PrivateMounts = true; # OK
};
```

Enabling any *single* one of the following triggered the original issue:

```
NoNewPrivileges = true; 
PrivateDevices = true;
ProtectHostname = true;
ProtectKernelTunables = true;
ProtectKernelModules = true;
RestrictAddressFamilies = [ "AF_UNIX" "AF_INET" "AF_INET6" ];
LockPersonality = true; 
MemoryDenyWriteExecute = !(builtins.any (mod: (mod.allowMemoryWriteExecute or false)) (optionals (cfg.package ? modules) cfg.package.modules));
RestrictRealtime = true; 
RestrictSUIDSGID = true;
SystemCallArchitectures = "native";
```

So, quite a lot of options that would need to be disabled..

**Comment by jluttine** [comment]
Oh, it's probably just about `NoNewPrivileges` basically. Those other settings just override it to `true` even if not set (or set to `false`): https://www.freedesktop.org/software/systemd/man/systemd.exec.html#NoNewPrivileges=

Is it possible to do PAM authentication without new privileges? :thinking:

**Comment by jluttine** [comment]
So, a workaround is to add the following to `configuration.nix`:

```
  systemd.services.nginx.serviceConfig = {
    NoNewPrivileges = lib.mkForce false;
    PrivateDevices = lib.mkForce false;
    ProtectHostname = lib.mkForce false;
    ProtectKernelTunables = lib.mkForce false;
    ProtectKernelModules = lib.mkForce false;
    RestrictAddressFamilies = lib.mkForce [];
    LockPersonality = lib.mkForce false;
    MemoryDenyWriteExecute = lib.mkForce false;
    RestrictRealtime = lib.mkForce false;
    RestrictSUIDSGID = lib.mkForce false;
    SystemCallArchitectures = lib.mkForce "";
  };
```

I don't know about the security implications and whether there is a better solution. So, I'm not closing this issue now.

**Comment by aanderse** [comment]
> I don't know about the security implications and whether there is a better solution. So, I'm not closing this issue now.

The security implications of such changes is that you would be running `nginx` the exact same way everyone else ran (and still runs)  `nginx` for **years** before `systemd` came along, or today on any OS/distro which doesn't have `systemd`.

`systemd` hardening is a very nice additional layer of security, but keep the word "additional" in mind here. Plenty of BSD servers happily running `nginx` without `systemd` hardening, so feel free to disable those options on your own server(s) if PAM helps you out. We might want to consider adding some functionality configure `nginx` to use PAM automatically and disable those additional hardening flags too... or maybe just a wiki article. :man_shrugging:

**Comment by jluttine** [comment]
Thanks! So, there's no way to have PAM authentication with nginx without disabling those security hardening options? If there is, that'd be great. In any case, PAM option for nginx service in NixOS would be nice. I could make a PR about it if I just know what's the preferred way of doing it. Not sure my workaround was the best option.

**Comment by aanderse** [comment]
Understood. Pull in a few of the `nginx` maintainers/contributors here and I'm sure some great ideas will be presented. Usually @Izorkin has a few tricks up their sleave :smile:

**Comment by Izorkin** [comment]
With sandbox mode error running `unix_chkpwd`
```
systemd-coredump[72542]: [🡕] Process 72540 (unix_chkpwd) of user 60 dumped core.
```

Need to find a running `unix_chkpwd` with `strace` - `strace -ff -o /var/log/unix_chkpwd_log /run/wrappers/bin/unix_chkpwd`

**Comment by DavHau** [comment]
I tried getting this to work, but by now it doesn't even seem to work anymore even when disabling all the hardening flags. I'm still getting:
```
Jul 28 17:05:37 grmpf-nix [648252]: pam_unix(nginx:auth): authentication failure; logname= uid=60 euid=60 tty= ruser= rhost=127.0.0.1  user=grmpf
Jul 28 17:05:39 grmpf-nix nginx[648252]: 2024/07/28 17:05:37 [error] 648252#648252: *1 PAM: user 'grmpf' - not authenticated: Authentication failure, client: 127.0.0.1, server: localhost, request: "GET / HTTP/1.1", host: "localhost"
```

At the same tie, this works fine: `sudo -u nginx pamtester nginx grmpf authenticate`

**Comment by mfrischknecht** [comment]
TL;DR: I have a way to reproduce the issue with `pamtester`, but I don't have a solution. I'd be glad about any input from people who know more PAM and systemd Voodoo than me :)



Like @DavHau, I also just ran into this issue. I used the [workaround ](https://github.com/NixOS/nixpkgs/issues/117578#issuecomment-808150962) mentioned by @jluttine (which worked a while ago), but that broke for me, too.

Now, I don't know enough about how PAM actually operates on a nitty gritty details level (and/or what changed in either Systemd or the NixOS integration that broke the NGINX-Auth further), and the things I tried to fiddle with didn't work out at all. I have, however, succeeded in adapting @DavHau's test command into something that actually reproduces the problem (i.e. the failure) on my machine. For that, have a look at this script: https://gist.github.com/mfrischknecht/79fbd792109428b580321c4aeb7e6799

In order to directly run the test, make sure that `pamtester` is in `environment.systemPackages` (including it in the command turned out to be way too complicated) and run `sudo ./test.sh run nginx <pam_service> <user>`.

For me, for example, the output looks something like this (obviously, behind `Password:`, I had to type in my password):

```
[manuel@apps:~/source/nginx-pam]$ sudo ./test.sh run nginx nginx_manuel manuel
Running as unit: test-pamtester.service; invocation ID: 1a9e4988f4454a7e915356c57d42f1a1
Press ^] three times within 1s to disconnect TTY.
Password:
pamtester: Authentication failure
Finished with result: exit-code
Main processes terminated with: code=exited, status=1/FAILURE
Service runtime: 3.927s
CPU time consumed: 29ms
Memory peak: 1.9M

Log:
Jan 02 05:42:42 apps systemd[1]: Started /run/current-system/sw/bin/pamtester nginx_manuel manuel authenticate.
Jan 02 05:42:42 apps pamtester[3051719]: pam_succeed_if(nginx_manuel:auth): requirement "user = manuel" was met by user "manuel"
Jan 02 05:42:42 apps pamtester[3051719]: pam_unix(nginx_manuel:auth): username [manuel] obtained
Jan 02 05:42:44 apps pamtester[3051719]: pam_unix(nginx_manuel:auth): authentication failure; logname= uid=60 euid=60 tty= ruser= rhost=  user=manuel
Jan 02 05:42:46 apps systemd[1]: test-pamtester.service: Main process exited, code=exited, status=1/FAILURE
Jan 02 05:42:46 apps systemd[1]: test-pamtester.service: Failed with result 'exit-code'.
```

Switching out the `run` command into `print`, the script simply prints out the underlying `systemd-run` CLI command that simulates the PAM auth attempt using the systemd properties of `nginx.service`.

---

Now, the interesting thing: For me, like for @DavHau, this succeeds:
```
[manuel@apps:~/source/nginx-pam]$ sudo -u nginx pamtester nginx_manuel manuel authenticate
Password:
pamtester: successfully authenticated
```

...while the `systemd-run` version doesn't. Except: If I'm running `./test.sh print`, copy the command, and remove the `--property=User=nginx` argument (i.e. I run the test as `root` instead of the `nginx` user), it works; so I assume the reason really is some restriction imposed by systemd, but I have no idea which one.

---

As for the systemd unit, I tried adding various combinations of the following additional settings, which also didn't help at all:
```
PrivateTmp = lib.mkForce false;
ReadWritePaths = lib.mkForce [ "/tmp" ];

ProtectSystem = lib.mkForce false;
PrivateUsers = lib.mkForce false;
ReadOnlyPaths = lib.mkForce [ "/etc/pam.d" "/etc/shadow" ];

CapabilityBoundingSet = lib.mkForce [ "CAP_DAC_READ_SEARCH" /* on top of: */ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
AmbientCapabilities   = lib.mkForce [ "CAP_DAC_READ_SEARCH" /* on top of: */ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
```

**Comment by mfrischknecht** [comment]
After some further experiments, I have an update on the problem @DavHau and I are facing:

After bisecting all possible capabilities, I found that these changes allow NGINX to use PAM again:

```
CapabilityBoundingSet = lib.mkForce [ "CAP_SETUID" /* on top of: */ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
AmbientCapabilities   = lib.mkForce [ "CAP_SETUID" /* on top of: */ "CAP_NET_BIND_SERVICE" "CAP_SYS_RESOURCE" ];
```

Now, I don't think this is a satisfactory solution, as it effectively gives `nginx` sudo rights, so I tried to find out why this should now suddenly be necessary. I traced `setuid` calls using `bpftrace`, and indeed, `nginx` attempts `setuid(0)` for each auth request:

```
[nix-shell:/home/manuel/source/nginx-pam]# cat trace.bt
#!/usr/bin/env bpftrace

tracepoint:syscalls:sys_enter_setuid
{
        $target_uid = args->uid;
        time("%H:%M:%S  ");
        printf("%-6d %-6d %-16s %-6d\n", uid, pid, comm, $target_uid);
        printf("%s\n\n", ustack);
}

[nix-shell:/home/manuel/source/nginx-pam]# ./trace.bt
Attaching 1 probe...
15:43:10  60     3109281 nginx            0

        0x7fe5feee4e08
```

---

After some searching on GitHub, I found this change here:
https://github.com/linux-pam/linux-pam/commit/b3020da7da384d769f27a8713257fbe1001878be

*If* I am reading this right, this change has altered the behavior of `pam_unix.so` such that it now _always_ attempts to verify passwords using a `unix_chkpwd` helper, which _always_ involves `setuid(0)` *UNLESS* the password check has been explicitly built into `pam_unix.so`; see here:
https://github.com/linux-pam/linux-pam/commit/b3020da7da384d769f27a8713257fbe1001878be#diff-d7ff6757d4f74b58323438695015c2f71e3d30c90033a14daae54e4b38faacafR254

...if `HELPER_COMPILE` is defined, then this code would not return `PAM_UNIX_RUN_HELPER`, but wind straight through to the bottom, returning `PAM_SUCCESS`, meaning that the password check would be attempted in the calling process directly, see:
https://github.com/linux-pam/linux-pam/blob/8c2bb459417f363bb6a8ad54081ba671d039a8f6/modules/pam_unix/support.c#L768

So, my understanding so far is that our problems stem mainly from the fact that `pam_unix.so` will attempt to run a password checking binary at all times now, unless we switch to a different build of PAM where `HELPER_COMPILE` has been set. I haven't figured out how the Nixpkgs build of PAM works exactly, though, so I don't know how easy this is to patch in.

**Comment by mfrischknecht** [comment]
TL;DR: Patching PAM seems like a bad idea, I went with another route. Try it out, if you like.

---

After some further investigation I now understand that the `HELPER_COMPILE` directive I mentioned above isn't meant as a build customization option; it looks like this is an internal directive that is used to differentiate the builds for `pam_unix.so` and the helper binaries (such as `unix_chkpwd`) at compile time -- they seem to share a large portion of their codebase otherwise.

This means that simply overriding some compiler flag during the build doesn't really work. I succeeded in patching `linux-pam` such that `pam_unix.so` would perform the checks itself (instead of running a helper), but in the end, that only means that `nginx` still runs in the `shadow` group, which I'm also not too much of a fan of.

After further thought and tests, I've now found a solution I'm considering good enough for my use case (which is not a very critical one: a personal home server not exposed to the internet). _Note_ that I'm neither an expert in security nor in how PAM operates (or its administration), so take this with a grain of salt. And: If someone here knows a better solution for these issues, I'm thankful for any pointers :)

---

The solution I arrived at was this: I wrote another [simple PAM helper binary](https://github.com/mfrischknecht/pam_proxy_helper/tree/master) that performs auth and account validity checks when called through `pam_exec.so`. This means I can have `nginx` issue its own PAM checks against a "proxy" policy which simply causes my helper to be called. For this `pam_proxy_helper`, I register a NixOS security wrapper that is `setuid(root)` and `setgid(shadow)`, which means that `nginx` itself now doesn't need to acquire either of these privileges anymore.

The upside of this is that my `nginx` service now also requires far fewer additional privileges than @jluttine [outlined in 2021](https://github.com/NixOS/nixpkgs/issues/117578#issuecomment-808150962); essentially, `nginx` only still needs the rights to run the binary (which implies `setuid` and `setgid`); but, importantly, `CAP_SETUID` now only has to be in the _bounding set_ of capabilities, _not in the ambient set_. Meaning: `nginx` itself shouldn't be able to call `setuid()` from its own logic.

```nix
{ config, lib, pkgs, ... }:

{
  systemd.services.nginx.serviceConfig = {
    NoNewPrivileges = lib.mkForce false;
    CapabilityBoundingSet = [ "CAP_SETUID" ];
    SystemCallFilter = [ "setuid" "setgid" "capset" ];
  };

  security.wrappers.pam_proxy_helper =
    let bin = (pkgs.callPackage ./pam_proxy_helper.nix { });
    in {
      owner = "root";   setuid = true;
      group = "shadow"; setgid = true;
      source = "${bin}/bin/pam_proxy_helper";
    };

  services.nginx = {
    enable = true;
    additionalModules = [ pkgs.nginxModules.pam ];

    virtualHosts."myapp.example.com" = {
      locations."/".proxyPass = "http://127.0.0.1:1234/";
      extraConfig = ''
        auth_pam "Restricted";
        auth_pam_service_name "nginx_proxy";
      '';
    };
  };

  security.pam.services = {
    nginx_proxy = {
      setEnvironment = false;
      text = ''
        auth required pam_exec.so expose_authtok ${config.security.wrapperDir}/pam_proxy_helper nginx_target
        account required pam_permit.so
        session required pam_permit.so
      '';
    };

    nginx_target = {
      setEnvironment = false;
      text = ''
        auth required pam_unix.so
        account required pam_unix.so
        session required pam_permit.so
      '';
    };
  };
}
```

For the helper binary, I use this `pam_proxy_helper.nix`:

```nix
{ lib
, rustPlatform
, fetchFromGitHub
, llvmPackages
, stdenv
, pam
}:

with import <nixpkgs> {};

rustPlatform.buildRustPackage rec {
  pname = "pam_proxy_helper";
  version = "0.1.0";

  src = fetchFromGitHub {
    owner = "mfrischknecht";
    repo = "pam_proxy_helper";
    rev = "versions/${version}";
    hash = "sha256-yRFsxUx0TUYYJo1HQlcJ7VI5YLmkduyY0SLQgDXUQ7I=";
  };

  cargoHash = "sha256-yYeL0iYdT007BYme4Nrao7A+W6SZB2/C2iN4hfvGzeE=";

  buildInputs = [ stdenv.cc.libc pam ];
  nativeBuildInputs = [ rustPlatform.bindgenHook ];
  LIBCLANG_PATH="${llvmPackages.libclang.lib}/lib";

  meta = {
    license = lib.licenses.agpl3Only;
    homepage = "https://github.com/mfrischknecht/pam_proxy_helper";
    description = "A proxy helper to run PAM auth requests through `pam_exec`";
  };
}
```
